command/agent/monitor/monitor.go

```diff
@@ -0,0 +1,76 @@
+package monitor
+
+import (
+	"sync"
+
+	log "github.com/hashicorp/go-hclog"
+)
+
+type Monitor struct {
+	sync.Mutex
+	sink         log.SinkAdapter
+	logger       log.InterceptLogger
+	logCh        chan []byte
+	index        int
+	droppedCount int
+	bufSize      int
+}
+
+func New(buf int, logger log.InterceptLogger, opts *log.LoggerOptions) *Monitor {
+	sw := &Monitor{
+		logger:  logger,
+		logCh:   make(chan []byte, buf),
+		index:   0,
+		bufSize: buf,
+	}
+
+	opts.Output = sw
+	sink := log.NewSinkAdapter(opts)
+	sw.sink = sink
+
+	return sw
+}
+
+func (d *Monitor) Start(stopCh <-chan struct{}) <-chan []byte {
+	d.logger.RegisterSink(d.sink)
+
+	logCh := make(chan []byte, d.bufSize)
+	go func() {
+		for {
+			select {
+			case log := <-d.logCh:
+				logCh <- log
+			case <-stopCh:
+				d.Lock()
+				defer d.Unlock()
+
+				d.logger.DeregisterSink(d.sink)
+				close(d.logCh)
+				return
+			}
+		}
+	}()
+
+	return logCh
+}
+
+// Write attempts to send latest log to logCh
+// it drops the log if channel is unavailable to receive
+func (d *Monitor) Write(p []byte) (n int, err error) {
+	d.Lock()
+	defer d.Unlock()
+
+	bytes := make([]byte, len(p))
+	copy(bytes, p)
+
+	select {
+	case d.logCh <- bytes:
+	default:
+		d.droppedCount++
+		if d.droppedCount > 10 {
+			d.logger.Warn("Monitor dropped %d logs during monitor request", d.droppedCount)
```

Comment:
Doesn't this end up calling d.Write and therefore deadlocking on trying to re-acquire the lock?

Another approach would be to have Start create another goroutine that periodically (every 3s?) monitored for dropped messages on a loop and sent along log lines like:

```go
dropped := fmt.Sprintf("...")
select {
case d.logCh <- dropped:
default:
  // Make room for dropped message
  select {
  case <-d.logCh:
    d.droppedCount++
    dropped = fmt.Sprintf("...")
  default:
  }
  d.logCh<-dropped
}
d.droppedCount = 0
```

As long as make an effort to report dropped lines in a way that doesn't risk causing more load on the slow downstream system, I'm fine with it. We can always revisit this if people end up getting hit hard with dropped lines.